Wikipedia/Code/NotificationsCenterViewModel.swift

```diff
@@ -25,57 +37,95 @@ final class NotificationsCenterViewModel: NSObject {
+    
+    func fetchFirstPage() {
+        kickoffImportIfNeeded { [weak self] in
+            
+            DispatchQueue.main.async {
+                guard let self = self else {
+                    return
+                }
+                
+                let notifications = self.remoteNotificationsController.fetchNotifications()
+                self.modelController.addNewCellViewModelsWith(notifications: notifications, editMode: self.editMode)
+                self.delegate?.cellViewModelsDidChange(cellViewModels: self.modelController.sortedCellViewModels)
+            }
+        }
     }
-
-    func cellViewModel(indexPath: IndexPath) -> NotificationsCenterCellViewModel? {
+    
+    func fetchNextPage() {
         
-        if let remoteNotification =  fetchedResultsController?.object(at: indexPath) {
-            return NotificationsCenterCellViewModel(notification: remoteNotification, languageLinkController: languageLinkController)
+        guard isPagingEnabled == true else {
+            DDLogDebug("Request to fetch next page while paging is disabled. Ignoring.")
+            return
         }
-
-        return nil
+        
+        let notifications = self.remoteNotificationsController.fetchNotifications(fetchOffset: modelController.fetchOffset)
+        
+        guard notifications.count > 0 else {
+            isPagingEnabled = false
+            return
+        }
+        
+        modelController.addNewCellViewModelsWith(notifications: notifications, editMode: self.editMode)
+        self.delegate?.cellViewModelsDidChange(cellViewModels: modelController.sortedCellViewModels)
+    }
+    
+    func toggleCheckedStatus(cellViewModel: NotificationsCenterCellViewModel) {
+        cellViewModel.toggleCheckedStatus()
+        reloadCellWithViewModelIfNeeded(viewModel: cellViewModel)
     }
 
 }
 
-extension NotificationsCenterViewModel: CollectionViewUpdaterDelegate {
-
-    func collectionViewUpdater<T>(_ updater: CollectionViewUpdater<T>, didUpdate collectionView: UICollectionView) where T : NSFetchRequestResult {
-        delegate?.collectionViewUpdaterDidUpdate()
+private extension NotificationsCenterViewModel {
+    func kickoffImportIfNeeded(completion: @escaping () -> Void) {
+        remoteNotificationsController.importNotificationsIfNeeded() { [weak self] error in
+            
+            guard let self = self else {
+                return
+            }
+            
+            if let error = error,
+               error == RemoteNotificationsOperationsError.failureSettingUpModelController {
+                //TODO: trigger error state of some sort
+                completion()
+                return
+            }
+            
+            NotificationCenter.default.addObserver(self, selector: #selector(self.contextObjectsDidChange(_:)), name: Notification.Name.NSManagedObjectContextObjectsDidChange, object: self.remoteNotificationsController.viewContext)
```

Comment:
Would it make sense for this observer subscription to just exist outside of this block, at the top of this function?

Reply:
@staykids actually this location was deliberate - in an earlier version I had this listener set up in the `init` statement of this view model even. The problem was as things are importing (particularly a large set of notifications), the `contextObjectsDidChange` handler starts firing and data is displayed before importing is completed. So the importing state goes away quickly and we start to see cells flying in and reordering for a while. Also because the main thread is so active inserting cells, it's practically unscrollable, so it's of little use to the user to see this data while it's importing. By only adding the listener here in this completion block, I can be sure languages are fully imported and the data is a bit more stable before `contextObjectsDidChange` fires.